src/java/org/apache/cassandra/db/compaction/unified/Controller.java

```diff
@@ -0,0 +1,723 @@
+     * dataset size that should be additionally available for compaction's space amplification overhead.
+     */
+    public double getMaxSpaceOverhead()
+    {
+        return maxSpaceOverhead;
+    }
+
+    /**
+     * @return whether is allowed to drop expired SSTables without checking if partition keys appear in other SSTables.
+     * Same behavior as in TWCS.
+     */
+    public boolean getIgnoreOverlapsInExpirationCheck()
+    {
+        return ignoreOverlapsInExpirationCheck;
+    }
+
+    public long getExpiredSSTableCheckFrequency()
+    {
+        return expiredSSTableCheckFrequency;
+    }
+
+    /**
+     * Perform any initialization that requires the strategy.
+     */
+    public void startup(UnifiedCompactionStrategy strategy, ScheduledExecutorService executorService)
+    {
+        if (calculator != null)
+            throw new IllegalStateException("Already started");
+
+        startup(strategy, new CostsCalculator(env, strategy, executorService, survivalFactor));
+    }
+
+    @VisibleForTesting
+    void startup(UnifiedCompactionStrategy strategy, CostsCalculator calculator)
+    {
+        this.calculator = calculator;
+        metrics = allMetrics.computeIfAbsent(strategy.getMetadata(), Controller.Metrics::new);
+        metrics.add(this);
+        logger.debug("Started compaction controller {}", this);
+    }
+
+    /**
+     * Signals that the strategy is about to be deleted or stopped.
+     */
+    public void shutdown()
+    {
+        if (calculator == null)
+            return;
+
+        calculator.close();
+        calculator = null;
+
+        if (metrics != null)
+        {
+            metrics.remove(this);
+            metrics = null;
+        }
+
+        logger.debug("Stopped compaction controller {}", this);
+    }
+
+    /**
+     * @return true if the controller is running
+     */
+    public boolean isRunning()
+    {
+        return calculator != null;
+    }
+
+    /**
+     * @return the cost calculator, will be null until {@link this#startup(UnifiedCompactionStrategy, ScheduledExecutorService)} is called.
+     */
+    @Nullable
+    @VisibleForTesting
+    public CostsCalculator getCalculator()
+    {
+        return calculator;
+    }
+
+    /**
+     * The strategy will call this method each time {@link CompactionStrategy#getNextBackgroundTasks(int)} is called.
+     */
+    public void onStrategyBackgroundTaskRequest()
+    {
+    }
+
+    /**
+     * Calculate the read amplification assuming a single scaling factor W and a given total
+     * length of data on disk.
+     *
+     * @param length the total length on disk
+     * @param W the scaling factor to use for the calculation
+     *
+     * @return the read amplification of all the buckets needed to cover the total length
+     */
+    public int RA(long length, int W)
+    {
+        double o = getSurvivalFactor();
+        long m = getMinSstableSizeBytes();
```

Comment:
These should use the flush size.

Reply:
Fixed in 58610d47730f805e2d023bda0748d79973918115.